### PR TITLE
feat: able to change the ordering of popup menu

### PR DESCRIPTION
**Pull request by kiddos**
hi.

I was wondering if we can have a feature to change the ordering of the popup menu. The current popup menu is ordered as "word", "kind", "extra". I'm using an option to control this for now, not sure if there's a better way to accomplish this.

https://github.com/vim/vim/assets/9444583/e4c7fbe4-be0a-4fa2-9757-2466d568e301

### CODE REVIEW

**Comment by chrisbra** [comment]
thanks, that sounds interesting. However I think the option value should not be an opaque integer value `012` but rather something like:  `:set pumordering=word,kind,extra` 

Hopefully that is flexible enough and others don't want printf() formatting strings (or add arbitrary items to the pum menu).

**Comment by kiddos** [comment]
I see that some of the computation can be done at the option set callback, do I need to move the computation to the callback?

**Comment by chrisbra** [comment]
I don't think so. The option callback is for completing option values. Does that work?

**Comment by chrisbra** [comment]
can you please fix the failing CI?

**Comment by kiddos** [comment]
I might need to add some tests too.

**Comment by chrisbra** [comment]
yes, please add tests

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -6273,6 +6273,23 @@ A jump table for the options with a short description can be found at |Q_op|.
 	Insert mode completion.  When zero as much space as available is used.
 	|ins-completion-menu|.
 
+						*'pumordering'* *'po'*
+'pumordering' 'po'	string (default "012")
```
Comment (chrisbra) [inline runtime/doc/options.txt]: this should be "word,kind,extra" ?

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -6273,6 +6273,23 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+	This option specify how the popup menu should look like
+
+	The default value:
+	0: word
+	1: kind
+	2: extra
```
Comment (chrisbra) [inline runtime/doc/options.txt]: same here, default value is no longer 0,1,2

File: src/popupmenu.c [hunk 3]
```diff
@@ -450,6 +510,27 @@ pum_redraw(void)
...
+    prefix_pum_widths[0] = pum_base_width;
+    prefix_pum_widths[1] = pum_kind_width;
+    prefix_pum_widths[2] = pum_extra_width;
+
+    pum_set_ordering(rounds);
+    // precompute widths
+    for (r = 0; r < 3; ++r)
+    {
+	if (rounds[r] == 0 && prefix_pum_widths[rounds[r]] > 0)
+	{
```
Comment (chrisbra) [inline src/popupmenu.c]: please remove braces around single-line statements

**Comment by chrisbra** [comment]
can you add a few tests for:
- `set pumordering=` (empty)
- `set pumordering=word,kind`
- `set pumordering=word`
- `set pumordering=extra`
- `set pumordering=does-not-exists`

**Comment by kiddos** [comment]
I would say the current behavior for this case `set pumordering=extra` is kinda weird.

I could make it so that `word`, and `kind` would also appear?

File: runtime/doc/options.txt (hunk 1 above)
Comment (kiddos) [reply runtime/doc/options.txt]: yeah

File: runtime/doc/options.txt (hunk 2 above)
Comment (kiddos) [reply runtime/doc/options.txt]: yea

File: src/popupmenu.c (hunk 3 above)
Comment (kiddos) [reply src/popupmenu.c]: sure thing

File: src/testdir/test_popup.vim [hunk 4]
```diff
@@ -1333,4 +1333,119 @@ func Test_pum_highlights_custom()
...
+    endfunc
+    set completeopt=menu
+    set completefunc=CompleteFunc
+    set pumordering=word,kind,extra
+  END
+  call writefile(lines, 'Xscript', 'D')
+  let buf = RunVimInTerminal('-S Xscript', {})
+  call TermWait(buf)
+  call term_sendkeys(buf, "iaw\<C-X>\<C-u>")
+  call TermWait(buf, 50)
```
Comment (zeertzjq) [inline src/testdir/test_popup.vim:1358]: `TermWait()` is not needed before `VerifyScreenDump()`

File: runtime/doc/options.txt [hunk 5]
```diff
@@ -6273,6 +6273,20 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+						*'pumordering'* *'po'*
+'pumordering' 'po'	string (default "word,kind,extra")
+	This option specify how the popup menu should look like
+
+	The default value: word,kind,extra
+
+	So the popup menu should look like the following
+	"word"  "kind"  "extra"
+
+	If this value is set to 102, the popup menu would look like this:
```
Comment (chrisbra) [inline runtime/doc/options.txt]: still reference to `102`

**Comment by chrisbra** [comment]
> I would say the current behavior for this case `set pumordering=extra` is kinda weird.

What do you mean with it?

> I could make it so that `word`, and `kind` would also appear?

So that would basically mean, we would be able to re-order items, but cannot leave out items, right?

**Comment by kiddos** [comment]
If user do this `set pumordering=extra`, only the `extra` will show up.
Yeah, I kind of want to make it so all items cannot be left out.

File: runtime/doc/options.txt (hunk 5 above)
Comment (kiddos) [reply runtime/doc/options.txt]: ooops

**Comment by chrisbra** [comment]
> If user do this set pumordering=extra, only the extra will show up.

Yes, it might be weird, and it sounds like a "don't complain if you use this setting and Vim does what you tell it. Don't do it then".

Not sure if this should be allowed to have  only 'extra' in 'pumordering'. Opinions?

**Pull request closed without merging**